src/main/java/org/apache/pulsar/io/jcloud/format/JsonFormat.java

```diff
@@ -171,4 +177,42 @@ private void putRecordEntry(Map<String, Object> jsonKeyValue, String key,
             jsonKeyValue.put(key, convertRecordToObject((GenericRecord) record, schema));
         }
     }
+
+    /**
+     * This method will try use TYPEREF and ARRAY_TYPEREF to read json value.
+     *
+     * Once the read is successful, the same type is used for the next read.
+     *
+     * @param jsonParser
+     * @return
+     * @throws IOException
+     */
+    private static Map<String, Object> dynamicReadValue(JsonParser jsonParser) throws IOException {
+        if (priorityTryTyperef == TYPEREF) {
+            try {
+                return JSON_MAPPER.get().readValue(jsonParser, TYPEREF);
+            } catch (MismatchedInputException e) {
+                log.info("Use Map<String, Object> read json failed, try to use List<Map<String, Object>>");
+                priorityTryTyperef = ARRAY_TYPEREF;
+                return readValueForArrayType(jsonParser);
+            }
+        } else {
+            try {
+                return readValueForArrayType(jsonParser);
+            } catch (MismatchedInputException e) {
+                log.info("Use List<Map<String, Object>> read json failed, try to use Map<String, Object>");
+                priorityTryTyperef = TYPEREF;
+                return JSON_MAPPER.get().readValue(jsonParser, TYPEREF);
+            }
+        }
+    }
+
+    private static Map<String, Object> readValueForArrayType(JsonParser jsonParser) throws IOException {
```

Comment:
I think we should use non-static methods here. The `dynamicReadValue` is stateful. `priorityTryTyperef` should also be non-static.

Reply:
Make sense, Fixed.